### PR TITLE
Marker attributes

### LINKED ISSUE #260: Marker attributes
As this project is growing in functionality, there are some places where Rust is not expressive enough to generate quality Dart code.
The idea is that bridging libraries e.g. wasm_bindgen also use attributes to guide code generation, for example in this case `#[wasm_bindgen(...)]`.

A simple use-case for our library would be declaring non-final fields: today, a struct in Rust corresponds to a class in Dart where
all fields are final with no way to configure. For example, this code in Rust:

```rust
struct Foo {
  pub foo: String
  pub bar: Option<String>
}
```

would be translated into this Dart code:

```dart
class Foo {
  final String foo;
  final String? bar;
  Foo({
    required this.foo,
    this.bar
  });
}
```

It would be impossible to update Foo.bar if it's null once it's sent to Dart, which might be undesirable for some usages.

With an attribute e.g. `#[frb]`, we can signal that some field actually needs to be non-final in Dart:

```rust
#[frb]
struct Foo {
  pub foo: String
  #[frb(mut)]
  pub bar: Option<String>
}
```

so that this Dart code is generated:

```dart
class Foo {
  final String foo;
  String? bar;
  Foo({
    required this.foo,
    this.bar
  });
}
```

These attributes will be beneficial for other features in the future as well, so there are many advantages to implementing attributes first.

### PR DESCRIPTION
**Pull request by Desdaemon**
Resolves #260.

Only the attribute macro itself is implemented, not any mechanism to retrieve information from the attributes.

Some bike-shedding:
- [x] Name of the attribute, currently `frb`.
- [x] Should it be visible to users by default? (currently **no**). This boils down to convenience vs. compile times, due to the added *compile-time* dependency on `syn`.

### CODE REVIEW

**Comment by fzyzcjy** [comment]
>  Name of the attribute, currently frb.

Agree ;)

> Should it be visible to users by default? (currently yes). This boils down to convenience vs. compile times, due to the added dependency on syn.

Why do we add dependency on syn? Since the proc macro itself does *nothing* in it, what about implementing a "raw" proc macro?

File: frb_example/pure_dart/dart/lib/bridge_generated.dart [hunk 1]
```diff
@@ -84,6 +84,8 @@ abstract class FlutterRustBridgeExample extends FlutterRustBridgeBase<FlutterRus
   Future<Foobar> handleEnumStruct({required Foobar val, dynamic hint});
 
   Future<String> handleComplexEnum({required KitchenSink val, dynamic hint});
+
+  Future<void> handleCustomiedStruct({required Customized val, dynamic hint});
```
Comment (fzyzcjy) [inline frb_example/pure_dart/dart/lib/bridge_generated.dart:88]: typo

**Comment by fzyzcjy** [comment]
What about this: let the `frb` macro does **nothing** inside that proc macro function body. Instead, **simply output the input**. Then no need for syn, and that `fn frb() {}` will almost be empty function.

**Comment by Desdaemon** [comment]
> Why do we add dependency on syn? Since the proc macro itself does nothing in it, what about implementing a "raw" proc macro?

It's needed to remove the attributes from the AST, as rustc does not accept attribute macros on fields. (Only derive macros are allowed to have field attributes.)

**Comment by Desdaemon** [comment]
> and that fn frb() {} will almost be empty function.

Also yes, the macro does zero work if it's not used on a struct/enum/union.

**Comment by fzyzcjy** [comment]
Hmm interesting... Can we do it without syn dependency?

**Comment by Desdaemon** [comment]
> Hmm interesting... Can we do it without syn dependency?

I've never tried this, it does seem a lot more effort upfront though. What we *could* do is re-implement a subset of syn's features, but of course we'd still be depending on the parsing functionality of syn which is only a fraction of the code.

**Comment by fzyzcjy** [comment]
> > Hmm interesting... Can we do it without syn dependency?
> 
> I've never tried this, it does seem a lot more effort upfront though. What we _could_ do is re-implement a subset of syn's features, but of course we'd still be depending on the parsing functionality of syn which is only a fraction of the code.

Interesting... Then we should not do that.

> Should it be visible to users by default? (currently yes). 

Then this becomes a question

**Comment by Desdaemon** [comment]
Related: rust-lang/rust#65823

**Comment by Desdaemon** [comment]
Given that this is a new feature with no real use cases yet, I think we can keep it gated behind a feature for now (already done), and we can reassess this need later on.

**Comment by fzyzcjy** [comment]
btw is there a crate that we can use directly, instead of manually writing down those calls to remove_marker_attr? seems like this should be a commonly used situation

**Comment by fzyzcjy** [comment]
going to merge it and publish it and change CI soon

**Comment by Desdaemon** [comment]
FWIW the last commit shaved off about 2 seconds on a debug clean build, so I'm hopeful it's not too much of an impact.

**Comment by Desdaemon** [comment]
> btw is there a crate that we can use directly, instead of manually writing down those calls to remove_marker_attr? seems like this should be a commonly used situation

I'm quite sure frb_codegen has no use for these, as they are only there to satisfy rustc? Anyways I can move them later if needed.

**Comment by fzyzcjy** [comment]
> I'm quite sure frb_codegen has no use for these, as they are only there to satisfy rustc? Anyways I can move them later if needed.

I mean these https://github.com/fzyzcjy/flutter_rust_bridge/blob/409acd9482c3478c4d1fe9607eba7da1ba3f7a15/frb_macros/src/lib.rs#L14-L72

is there someone already write down a crate called, e.g. "create_proc_macro_that_does_nothing" on crates.io and we can use that

**Comment by fzyzcjy** [comment]
> FWIW the last commit shaved off about 2 seconds on a debug clean build, so I'm hopeful it's not too much of an impact.

Agree. not a big problem

**Comment by Desdaemon** [comment]
> is there someone already write down a crate called, e.g. "create_proc_macro_that_does_nothing" on crates.io and we can use that

Strangely enough, I couldn't find anything like that, the closest thing was wasm_bindgen itself which uses more features of syn.

**Comment by fzyzcjy** [comment]
interesting. so let me merge it

I may not have time immediately and will publish tomorrow (ping me if I forget)

**Pull request merged**